### PR TITLE
Tentative refactor of SQLite.jl

### PR DESCRIPTION
**Pull request by metab0t**
1. Use Clang.jl to generate API bindings (a new C module).
2. Separate functions to different files for clarity. (I feel that monotholic code structure is not friendly for maintainers)
3. Use WeakKeyDict to record prepared statements of a DB, so that statements can be GCed independently from DB. (Currently SQLite.Stmt does not contain the pointer to sqlite3_stmt which is cubersome)
4. Only ONE public API change: Change `SQLite.execute` to `SQLite.direct_execute` to avoid name confusion with `DBInterface.execute`.

The tests are almost unchanged and all passed on my machine.

Should fix https://github.com/JuliaDatabases/SQLite.jl/issues/278 and provides a better basis for https://github.com/JuliaDatabases/SQLite.jl/issues/284.

I can separate it into smaller PRs if it is needed.

### CODE REVIEW

**Comment by metab0t** [comment]
For `SQLite.execute` we can add a fallback to `SQLite.direct_execute` so that exposed API is nearly not affected.

File: test/runtests.jl [hunk 1]
```diff
@@ -42,7 +42,7 @@ Copy test sqlite file to temp directory, path `test.sqlite`,
 overwrite file if necessary and set permissions.
 """
 function setup_clean_test_db(f::Function, args...)
-    dbfile = joinpath(dirname(pathof(SQLite)), "../test/Chinook_Sqlite.sqlite")
+    dbfile = joinpath(@__DIR__, "Chinook_Sqlite.sqlite")
```
Comment (quinnj) [inline test/runtests.jl:45]: can we revert this change? I like the `pathof` style since it makes it much easier for running this locally, step-wise, in the REPL.

File: src/SQLite.jl [hunk 2]
```diff
@@ -1,726 +1,25 @@
...
-    @CHECK db sqlite3_prepare_v2(db.handle, sql, stmt, null)
-
+import Random
+import DBInterface
+using Serialization
+using WeakRefStrings
+using Tables
+
+include("capi.jl")
+import .C as C
```
Comment (quinnj) [inline src/SQLite.jl:10]: I think this can just be `import .C` since the `as` isn't renaming it at all. This is also causing test failures for Julia 1.3 (`as` didn't exist then). I'm also fine bumping the Julia version up to 1.6 though.

**Comment by quinnj** [comment]
This looks great! Thanks for doing this @metab0t! I really like the auto-generated C bindings. In the future, just as an FYI, it can make reviewing much easier to avoid reorganizing files in addition to lots of changes. I don't really mind the file reorganizing, but from a `git diff` perspective, it can be much easier to make all the proposed changes leaving files as-is, and then do the file reorganization in a separate PR.

**Comment by metab0t** [comment]
Thanks for reviewing!

My current plan is to separate it into 3 PRs:

1. C bindings and other internal changes in place without influencing API.
2. File reorganization
3. Rename or deprecate `SQLite.execute`

I will close this huge PR after I finish the separation.

**Comment by quinnj** [comment]
@metab0t , good to close this now?
